flag_engine/identities/models.py

```diff
@@ -1,28 +1,65 @@
 import datetime
 import typing
 import uuid
-from dataclasses import dataclass, field
 
+from pydantic import UUID4, BaseModel, Field, root_validator
+from pydantic_collections import BaseCollectionModel
+
+from flag_engine.features.models import FeatureStateModel
 from flag_engine.identities.traits.models import TraitModel
-from flag_engine.utils.collections import IdentityFeaturesList
 from flag_engine.utils.datetime import utcnow_with_tz
+from flag_engine.utils.exceptions import DuplicateFeatureState
+
+
+class IdentityFeaturesList(BaseCollectionModel[FeatureStateModel]):
+    @classmethod
+    def __get_validators__(
+        cls,
+    ) -> typing.Generator[typing.Callable[..., typing.Any], None, None]:
+        yield cls.validate
+        yield cls.ensure_unique_feature_ids
+
+    @classmethod
+    def ensure_unique_feature_ids(
+        cls, value: "IdentityFeaturesList"
+    ) -> "IdentityFeaturesList":
+        for i, feature_state in enumerate(value, start=1):
+            if feature_state.feature.id in [
+                feature_state.feature.id for feature_state in value[i:]
```

Comment:
Perhaps python will handle some optimisations for this internally but it certainly feels like we could optimise this by generating the list of feature ids up front?

Reply:
The point of those partial iterations are to avoid iterating over the full list in case duplicates are found. I lifted this approach from Pydantic's `ConstrainedList.unique_items_validator` implementation. Generating the feature id list would negate the optimisation. Maybe it's redundant for our data and we could simplify this code to iterate over all the ids right away and only do the membership check once.

Reply:
Ok, yeah, that makes sense. I guess, realistically this list is going to be a few items long in 99% of cases so we just shouldn't care about it either way :D